### PR TITLE
Decompile state system, dYesNoWindow_c and d_profile.cpp

### PR DESCRIPTION
**Pull request by RootCubed**
Additionally:

- Fixes a few build toolchain issues related to REL building:
  - Don't look up symbol by name in own elf, just check if `st_shndx != SHN_UNDEF`
  - Don't use `FORCEFILES`, use `EXCLUDEFILES` with `wiimj2d.elf` instead

- Adds `-ipa file` the build command-line options (needed for matching the state machine code), which meant fixing up some other already decompiled code.

- Slightly changes the profile creation macro; now uses a new enum for the draw order.

### CODE REVIEW

File: include/d_profileNP/bases/d_yes_no_window.hpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
+#pragma once
+#include <dol/bases/d_base.hpp>
+#include <dol/bases/d_lytbase.hpp>
+#include <dol/sLib/s_State.hpp>
+
+/// @brief Displays a yes/no window.
+/// @details This appears, for example, in the pause menu or when you press + on the world map.
```
Comment (CLF78) [inline include/d_profileNP/bases/d_yes_no_window.hpp]:
```suggestion
/// @brief Displays a confirmation prompt.
/// @details This appears, for example, when the game asks the player if their progress should be saved.
```

File: include/d_profileNP/bases/d_yes_no_window.hpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+/// @ingroup bases
+class dYesNoWindow_c : public dBase_c {
+public:
+    /// @brief The type of the window.
```
Comment (CLF78) [inline include/d_profileNP/bases/d_yes_no_window.hpp]:
```suggestion
    /// @brief The prompt type.
```

File: include/d_profileNP/bases/d_yes_no_window.hpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    };
+    
+    /// @brief The possible cursor positions.
+    enum CURSOR_POS_e {
+        POS_YES,
+        POS_NO,
+        POS_OK
+    };
+
+    /// @brief The sound effects used for the yes/no window.
```
Comment (CLF78) [inline include/d_profileNP/bases/d_yes_no_window.hpp]:
```suggestion
    /// @brief The sound effect types used by the window.
```

File: include/d_profileNP/bases/d_yes_no_window.hpp [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    enum SOUND_e {
+        SOUND_CLOSE,
+        SOUND_WAIT,
+        SOUND_CURSOR_MOVE,
+        SOUND_OPEN_ANIME
+    };
+
+    /// @brief The animations used for the yes/no window.
```
Comment (CLF78) [inline include/d_profileNP/bases/d_yes_no_window.hpp]:
```suggestion
    /// @brief The animations used for the window.
```

File: include/d_profileNP/bases/d_yes_no_window.hpp [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    virtual int execute();
+    virtual int draw();
+    virtual int doDelete();
+
+private:
+    /// @brief Fills in the layout of the window.
+    /// @details This includes:
+    /// - Filling out the text boxes
+    /// - Setting the visibility of the buttons
+    /// - Conditionally showing the quick save balloon
```
Comment (CLF78) [inline include/d_profileNP/bases/d_yes_no_window.hpp]:
```suggestion
    /// @brief Fills out the window content according to the settings.
    /// @details This includes:
    /// - Filling out the text boxes
    /// - Setting the visibility of the buttons
    /// - Showing the quick save balloon if necessary
```

File: include/d_profileNP/bases/d_yes_no_window.hpp [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    void populateLayout();
+
+    STATE_FUNC_DECLARE(dYesNoWindow_c, InitWait); ///< @brief Initial state. Switches to OpenAnimeStartWait immediately.
+    STATE_FUNC_DECLARE(dYesNoWindow_c, OpenAnimeEndWait); ///< @brief Waiting for the window to finish the opening animation.
+    STATE_FUNC_DECLARE(dYesNoWindow_c, ButtonOnStageAnimeEndWait); ///< @brief Playing the button to finish the appearing animation.
```
Comment (CLF78) [inline include/d_profileNP/bases/d_yes_no_window.hpp]:
```suggestion
    STATE_FUNC_DECLARE(dYesNoWindow_c, ButtonOnStageAnimeEndWait); ///< @brief Waiting for the button to finish the appearing animation.
```
That's my assumption of what this means at least :P

File: include/dol/bases/d_game_com.hpp [hunk 7]
```diff
@@ -8,4 +9,7 @@ namespace dGameCom {
     u32 getRandomSeed(); ///< Gets a seed used for randomness.
     void clearGameStop();
     float getDispCenterY();
+    
+    /// @unofficial
+    void updateSelectCursor(nw4r::lyt::Picture *pic, int index, bool unk);
```
Comment (CLF78) [inline include/dol/bases/d_game_com.hpp]:
```suggestion
    void updateSelectCursor(nw4r::lyt::Picture *pic, int index, bool useSpecialDraw);
```

File: include/dol/bases/d_s_world_map_static.hpp [hunk 8]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+#include <types.h>
+
+class dScWMap_c {
+public:
+    static u8 m_WorldNo;
+};
```
Comment (CLF78) [inline include/dol/bases/d_s_world_map_static.hpp]:
```suggestion
};

```
Missing ending newline.

File: include/dol/sLib/s_FStateID.hpp [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+#include <lib/MSL_C/string.h>
+
+/// @brief An implementation of a state ID for a given class.
+/// @details It adds the ability to call the three state methods on a state owner class.
+/// @tparam T The class that this state belongs to.
+/// @ingroup state
+template<class T>
+class sFStateID_c : public sStateID_c {
+public:
+    typedef void (T::*statePTMF)();
```
Comment (CLF78) [inline include/dol/sLib/s_FStateID.hpp]: Do we really need to call them PTMF? Could we call 'em "stateFunc" or something?

File: include/dol/sLib/s_StateInterfaces.hpp [hunk 10]
```diff
@@ -11,58 +11,65 @@ class sStateIDIf_c {
...
 public:
     virtual ~sStateFctIf_c() {}
-    virtual sStateIf_c *build(sStateIDIf_c const &id) = 0; ///< @pure Returns a new state with a given state ID.
-    virtual void dispose(sStateIf_c *&id) = 0; ///< @pure Clears out the pointer to a state.
+    virtual sStateIf_c *build(sStateIDIf_c const &id) = 0; ///< Returns a new state with a given state ID.
+    virtual void dispose(sStateIf_c *&id) = 0; ///< Clears out the pointer to a state.
+};
+
+/// @brief The interface for state ID checkers.
+/// @details [This class is not really used, but it seems to be intended for testing if a state is "normal", but not sure how this would be useful].
```
Comment (CLF78) [inline include/dol/sLib/s_StateInterfaces.hpp]:
```suggestion
/// @details [This class is not really used, but it seems to be intended for testing if a state is "normal", most likely a debug leftover].
```

File: source/d_profileNP/bases/d_yes_no_window.cpp [hunk 11]
```diff
@@ -0,0 +1,595 @@
...
+    mLayout.NPaneRegister(paneNames, &N_otehonText_00, ARRAY_SIZE(paneNames));
+    
+    mIsActive = false;
+    mHasLoadedLayout = true;
+    mKeepOpen = false;
+    mStarCoinsRequired = 0;
+    mCancel = false;
+    mHasBG = false;
+    
+    mpRootPane->mFlags &= ~1;
```
Comment (CLF78) [inline source/d_profileNP/bases/d_yes_no_window.cpp:129]: An enum for these will be needed at some point...

File: source/d_profileNP/bases/d_yes_no_window.cpp [hunk 12]
```diff
@@ -0,0 +1,595 @@
...
+        mLayout.entry();
+    }
+    return SUCCEEDED;
+}
+
+int dYesNoWindow_c::doDelete() {
+    return mLayout.doDelete();
+}
+
+/// @brief IDs for the descriptions in the Yes/No window.
```
Comment (CLF78) [inline source/d_profileNP/bases/d_yes_no_window.cpp]:
```suggestion
/// @brief The prompt message for each type.
```

File: source/d_profileNP/bases/d_yes_no_window.cpp [hunk 13]
```diff
@@ -0,0 +1,595 @@
...
+#include <dol/bases/d_audio.hpp>
+#include <dol/bases/d_game_com.hpp>
+#include <dol/bases/d_cursor_select.hpp>
+#include <dol/snd/snd_audio_mgr.hpp>
+#include <dol/bases/d_info.hpp>
+#include <dol/bases/d_s_world_map_static.hpp>
+#include <constants/message_list.h>
+#include <constants/game_constants.h>
+#include <constants/sound_list.h>
+
```
Comment (CLF78) [inline source/d_profileNP/bases/d_yes_no_window.cpp:14]:
```suggestion
/// @file

```
Docs added to this file will not be detected without this directive.

File: source/d_profileNP/bases/d_yes_no_window.cpp [hunk 14]
```diff
@@ -0,0 +1,595 @@
...
+STATE_DEFINE(dYesNoWindow_c, ButtonChangeAnimeEndWait);
+STATE_DEFINE(dYesNoWindow_c, SelectWait);
+STATE_DEFINE(dYesNoWindow_c, HitAnimeEndWait);
+STATE_DEFINE(dYesNoWindow_c, ClouseAnimeEndWait);
+STATE_DEFINE(dYesNoWindow_c, HitAnimeAfterWait);
+
+BASE_PROFILE(YES_NO_WINDOW, dYesNoWindow_c);
+
+dYesNoWindow_c::dYesNoWindow_c() : mStateMgr(*this, StateID_InitWait) {
+    mHasLoadedLayout = false;
```
Comment (CLF78) [inline source/d_profileNP/bases/d_yes_no_window.cpp]: No static initializer for the bool? _Insert megamind meme here_

File: source/d_profileNP/bases/d_yes_no_window.cpp (hunk 11 above)
Comment (RootCubed) [reply source/d_profileNP/bases/d_yes_no_window.cpp:129]: Yeah, we can do that when we decompile (or document more of) the nw4r::lyt stuff

**Pull request merged**